rubix-core/src/main/java/com/qubole/rubix/core/ReadRequestChainStats.java

```diff
@@ -12,174 +12,232 @@
  */
 package com.qubole.rubix.core;
 
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.ImmutableMap;
+import com.qubole.rubix.spi.BookKeeperFactory;
+import com.qubole.rubix.spi.RetryingPooledBookkeeperClient;
+import org.apache.hadoop.conf.Configuration;
+
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
 /**
  * Created by stagra on 28/1/16.
  */
 public class ReadRequestChainStats
 {
-  // All data below in MB
-  // From Remote
-  private long prefixRead;
-  private long suffixRead;
-  private long requestedRead;
-  private long warmupPenalty;
-  private long remoteReads;
+  public static final String DOWNLOADED_FOR_NON_LOCAL_METRIC = "downloaded_for_non_local";
+  public static final String EXTRA_READ_FOR_NON_LOCAL_METRIC = "extra_read_for_non_local";
+  public static final String WARMUP_TIME_NON_LOCAL_METRIC = "warmup_penalty_non_local";
+  public static final String DOWNLOADED_FOR_PARALLEL_WARMUP_METRIC = "downloaded_for_parallel_warmup";
+  public static final String PARALLEL_DOWNLOAD_TIME_METRIC = "parallel_download_time";
+
+  private static final long UNKNOWN_VALUE = -1;
+
+  // bookKeeperStats are fetched over thrift in standalone mode
+  // and are usually needed for multiple getters in this class
+  // cache the server response to minimize thrift calls but evict aggressively
+  private static final Cache<String, Map<String, Long>> bookKeeperStats = CacheBuilder
+          .newBuilder()
+          .expireAfterWrite(5, TimeUnit.SECONDS)
+          .build();
+  private static final String STATS_KEY = "stats";
+
+  // RemoteRRC
+  private long remoteRRCDataRead;
+  private long remoteRRCExtraDataRead;
+  private long remoteRRCWarmupTime;
+  private long remoteRRCRequests;
+
+  // CachedRRC
+  private long cachedRRCDataRead;
+  private long cachedRRCRequests;
+
+  // DirectRRC
+  private long directRRCDataRead;
+  private long directRRCRequests;
+
+  // NonLocalRRC
+  private long nonLocalRRCDataRead;
+  private long nonLocalRRCRequests;
 
-  // From cached
-  private long cachedDataRead;
-  private long cachedReads;
+  private int corruptedFileCount;
 
-  private long nonLocalReads;
-  private long nonLocalDataRead;
+  public ReadRequestChainStats add(ReadRequestChainStats other)
+  {
+    return new ReadRequestChainStats()
```

Comment:
Why not just update the current object instead of always creating a new one ?

Reply:
Good point! I don't remember why this was done originally but that is being carried over since. Looks safe to update in-place, will change this.

Reply:
done